source/common/upstream/thread_aware_lb_impl.cc

```diff
@@ -185,5 +186,100 @@ LoadBalancerPtr ThreadAwareLoadBalancerBase::LoadBalancerFactoryImpl::create() {
   return lb;
 }
 
+bool ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::isHostOverloaded(
+    HostConstSharedPtr host, double weight) const {
+
+  if (is_host_overloaded_ != nullptr) {
+    // this is only used while testing.
+    return is_host_overloaded_(host, weight);
+  }
+  /*
+  Consistent Hashing with Bounded Load
+  Ref: https://arxiv.org/abs/1608.01350
+  */
+  uint32_t overall_active = host->cluster().stats().upstream_rq_active_.value();
+  uint32_t host_active = host->stats().rq_active_.value();
+
+  uint32_t total_slots = ((overall_active + 1) * hash_balance_factor + 99) / 100;
+  uint32_t slots =
+      std::max(static_cast<uint32_t>(std::ceil(total_slots * weight)), static_cast<uint32_t>(1));
+
+  if (host->stats().rq_active_.value() > slots) {
+    ENVOY_LOG_MISC(
+        debug,
+        "ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::chooseHost: "
+        "host {} overloaded; overall_active {}, host_weight {}, host_active {} > slots {}",
+        host->address()->asString(), overall_active, weight, host_active, slots);
+    return true;
+  }
+  return false;
+}
+
+HostConstSharedPtr
+ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::chooseHost(uint64_t hash,
+                                                                        uint32_t attempt) const {
+
+  if (hlb_ptr == nullptr) {
+    return nullptr;
+  }
+  if (normalized_host_weights_ == nullptr || (*normalized_host_weights_).empty()) {
+    return nullptr;
+  }
+
+  HostConstSharedPtr host = hlb_ptr->chooseHost(hash, attempt);
+  if (host == nullptr) {
+    return nullptr;
+  }
+  const double weight = normalized_host_weights_map_.at(host);
+  if (!isHostOverloaded(host, weight)) {
+    ENVOY_LOG_MISC(debug,
+                   "ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::chooseHost: "
+                   "selected host #{} (attempt:1)",
+                   host->address()->asString());
+    return host;
+  }
+
+  /*
+  Revisiting Consistent Hashing with Bounded Loads
+  https://arxiv.org/abs/1908.08762
+
+  When a host is overloaded, we choose the next host in a random manner rather than picking the
+  next one in the ring. The random sequence is seeded by the hash, so the same input gets the same
+  sequence of hosts all the time.
+  */
+  uint32_t num_hosts = normalized_host_weights_->size();
+  auto host_index = std::vector<uint32_t>(num_hosts);
+  for (uint32_t i = 0; i < num_hosts; i++) {
+    host_index[i] = i;
+  }
+
+  uint64_t seed = hash;
+  std::default_random_engine random(seed);
```

Comment:
@zuercher there doesn't seem to be a way to set a seed on Random::RandomGenerator which is a requirement here. We need the same shuffle sequence for the same hash every time. We dont need a strong/secure random generator, the one from standard library should be good enough. Is there a specific reason you recommend using Random::RandomGenerator?

Reply:
Partly because it's the rng we use everywhere else, but also using `std::default_random_engine` gets a library-implementation defined RNG. So I think it'd be acceptable to use a specific engine here, and comment why RandomGenerator isn't being used.

Reply:
Thanks, I will add the comment. Seeded RNG is the main thing, std::default_random_engine can be made to use our own RNG, see: 
https://stackoverflow.com/questions/39615931/making-own-random-number-class-compitible-with-uniform-int-distribution